### DISCUSSION TITLE
Mass interactions and ways to improve jitter

### DISCUSSION THREAD

**Discussion by SSNTails** [jrouwe/JoltPhysics#1055] [Q&A] [answered]
Video #1: The player (dynamic sphere of mass 100) will push objects of a lower mass (convex hull dynamic crate with a mass of 1) through triangle meshes. Aren't static triangle meshes supposed to have infinite mass and therefore the crate should favor clipping through the player sphere instead? Or is there a way to allow the player to stand on top of a crate with a mass lower than the player?

https://github.com/jrouwe/JoltPhysics/assets/159929665/eac718bb-2d88-4e13-b1b5-c50c00ca9d5b

Video #2: This shows some barrels (dynamic convex hull) with a mass of 200, while the player is a dynamic sphere of 100 mass. The player can now land and stand on top of the barrels (as desired), but there's some jitter to it.. both the player and the barrel clip inside other objects (between themselves and the triangle mesh) and slowly push themselves into the correct position. Is there a way to reduce that artifacting?

https://github.com/jrouwe/JoltPhysics/assets/159929665/82df91b5-8554-402b-9a6e-b31aa969ab7c

**Comment by jrouwe** [comment]
Iterative solvers (which is what Jolt uses) are not very good in 'high mass ratios' (where the top object is much heavier than the bottom one). An object that's 100x more heavy than another object will not work (a factor 10x is already a lot). There is no 'priority system' that decides to push 1 object over pushing another object.

W.r.t. the barrel: I don't immediately see what could be going wrong. I'll need some sort of repro for that.

**Reply by SSNTails** [reply]
Ah, alright. I guess there's no need to have such large disparity in mass values. Here's a video again with the player with a mass of 1, and the crate and barrel with a mass of 1.1 and 1.2, respectively:

https://github.com/jrouwe/JoltPhysics/assets/159929665/988e5264-78f4-4dd4-b54d-eb7963dbfaa1

Then I thought, how do other shapes behave in this scenario? So I changed the crate into a Box shape (don't mind that the wireframe doesn't match up with the rendered crate). I know you said you needed a repro, but maybe this will spark some thought while I try to think of how to put one together:

https://github.com/jrouwe/JoltPhysics/assets/159929665/e9ffbd1c-f5dc-4e56-bf9b-b1a6567a880c

**Reply by jrouwe** [reply]
Sorry, I really can't tell what's wrong from a video.

One way you can try to create a repro is by saving a snapshot and loading it in the sample app's `LoadSnapshotTest`.

This is the code that is used to create a snapshot:

https://github.com/jrouwe/JoltPhysics/blob/c1ee107a476e7d7402bd5b09920bcfbafea42500/Samples/SamplesApp.cpp#L777-L788

**Reply by SSNTails** [reply]
Oh, that's easy and convenient! Here are a few snapshots:

ww14.bin) These convex hulls mysteriously fall through the triangle mesh. In my engine, they continuously stutter and act like the solver is trying to push them up to a resting place, but ultimately they fall through the floor.

[snapshots.zip](https://github.com/jrouwe/JoltPhysics/files/14984383/snapshots.zip)


standing & standing2.bin) Here are two variations on the videos above.

In all of these, why do the dynamic shapes fall through the static triangle mesh in the sample app? Is that a setting somewhere?

**Reply by jrouwe** [reply]
> In all of these, why do the dynamic shapes fall through the static triangle mesh in the sample app? Is that a setting somewhere?

If I load these scenes in the LoadSnapshotTest then I have to turn on the "Test Settings" / "Override Object Layers" (because you're using different physics layers than the sample app) and then it works reasonably well.

I get 2 asserts on startup coming from `CollisionDispatch::sInit` saying `Unsupported shape pair`. There are 2 pairs of `MeshShapes` that try to collide, this is not supported so no collisions will be returned there (this may be an artifact from the changed layers, the rule is really simple: all static goes in 1 layer, all dynamic in another).

standing.bin:

![image](https://github.com/jrouwe/JoltPhysics/assets/1621693/997c16fd-b390-4ee3-86d9-00a544d1b0a1)

This one seems to work reasonably well on my machine. The convex hull starts slightly penetrating (green arrow) but gets resolved within a second:

![image](https://github.com/jrouwe/JoltPhysics/assets/1621693/aaab1cb0-3e97-4433-b63c-be32f602c598)

standing2.bin:

This one starts with a very large penetration. Depenetrating a convex object vs a mesh (especially with a sphere on top that tries to push it back) is not an exact science and causes a lot of jitter:

![image](https://github.com/jrouwe/JoltPhysics/assets/1621693/d0d492c3-7e5d-4103-a562-7814a7f532be)

It's not a good idea to have 2 objects start with interpenetration that is this big (in principle you don't want any penetration to start with).

ww14.bin:

I'm not sure what you mean with 'these convex hulls', the only thing I see falling through the world is:

![image](https://github.com/jrouwe/JoltPhysics/assets/1621693/3c0d5185-a351-4ad9-af17-c2a1d5288c1d)

which just starts well below the triangle mesh (and the mesh is just an infinitely thin sheet, so if you're behind a triangle you're outside and will never get back).

And I see these sensors fall:

![image](https://github.com/jrouwe/JoltPhysics/assets/1621693/91cb6a4c-16f0-4455-94e4-e15b61c50f6a)

But sensors only detect collisions and will not be stopped by them.

**Reply by SSNTails** [reply]
Thank you for taking the time to look at this -- you're correct that mesh collisions were due to the layers not matching. I reworked the layers to match the sample app and it was much easier to experiment.
Some of those sample snapshots have deep penetrations because I took the snap after putting the objects in those compromising positions by an interaction of the player -- they aren't placed in the map like that.

I think the issue with being unable to repro it in the sample app is due to a few things:

1) In my app, mPhysicsSystem->SetGravity is 0, 0, 0. I apply gravity individually to each object myself as part of SetLinearVelocity() before each physics system update. This is because each object can have its own local gravity. I experimented with using the built-in gravity though, and didn't notice much behavioral change. Does the system do anything internally with the global gravity setting differently than SetLinearVelocity()? Is there a way I can mimick whatever this behavioral change is while keeping unique gravity per-object?

Because right now, even if the object is sitting on the ground, a gravitational force is being applied... I suspect this might be why the solver is struggling to 'push' the objects to a resting place. Adding the weight (and gravity velocity) of the player standing on top of one of these objects is just too much. So I guess I need to figure out a way to determine 'gravity no longer need to be applied here', which I assume is what Jolt is doing internally when you use the global gravity setting.

If you're wondering 'what's the use case?' imagine multiplayer game where you can sometimes walk on the ceiling. In a multiplayer situation (or if you don't want to affect other parts of the world), you don't want to flip the entire global gravity. Another use case is adding gravitational force based on the UP vector of the player the faster he moves, allowing better sticking to ramps, curved walls, etc.

2) If I change the # of iterations in mPhysicsSystem->Update to be greater than 1, things are MUCH more stable. But I also get some wiggling artifacts like so, but at least those large barrels aren't falling through the floor anymore!


https://github.com/jrouwe/JoltPhysics/assets/159929665/029060b6-451a-486e-9874-215b1b054a73

https://github.com/jrouwe/JoltPhysics/assets/159929665/d8a58274-621a-40da-a872-43866ad70e30


So TLDR; I think I need a new way to apply gravity to objects. Is there a way I can do this on a per-object basis in Jolt instead of globally?

**Reply by jrouwe** [reply]
If you want to apply individual gravity to each object then use `Body::AddForce` instead of `SetLinearVelocity`. Alternatively you can also tweak the gravity factor (`BodyCreationSettings::mGravityFactor`) for each object individually.

**Reply by jrouwe** [reply]
One more thing: Please take care that you stop applying forces / change velocities when the body is sleeping (`!Body::IsActive()`), unless you go through `BodyInterface::AddForce` etc. the body will not be woken up if you do it directly on the body. This means that it will accumulate a lot of force/speed and then when you finally wake the body up (e.g. by jumping on it) it will glitch because it has a massive speed for 1 frame.

**Reply by SSNTails** [reply]
Awesome!! I switched the gravity to using AddForce(), and now I get behavior very similar to that of the sample app! This is amazing! I hope I can showcase what I've been working on soon.

It does appear that the units used by AddForce() are different from SetLinearVelocity()'s meters/sec, though. The documentation says "Unit: N", what does that mean? Tried searching the docs for the keyword 'Unit', but it says zero results (however I know that can't possibly be true...)

**Reply by SSNTails** [reply]
Heh, nevermind, the main index says Newtons. I don't know what's up with the search. Is there a way I can convert m/s gravity to Newtons? I tried an online converter, but the value isn't accurate (the player mass is 1.0, so that shouldn't be a factor)

**Reply by jrouwe** [reply]
Force (Newton) = mass (kg) * acceleration (m/s^2) (and the acceleration = gravity in this case)

**Reply by SSNTails** [reply]
Adding the appropriate gravity back, I ended up with the same artifacts I originally had. But then I reduced the size of my worlds to 1/5th of their original size (thankfully, I have a global scalar that allows me to do this easily). Suddenly, all of the problems went away.

I'm not sure why this didn't express itself when loading the snapshot into the sample app. Maybe because the default gravity is lower in the sample app? (when I used lower gravity, I didn't get the same effects).

I then tried compiling Jolt using JPH_DOUBLE_PRECISION, with my original world unit size, and it still exhibited my original problematic behavior. So it's not a precision issue - at least one that JPH_DOUBLE_PRECISION can solve.

I don't have any question here.. just some information for you. For now, I'm using the smaller world units and will see if I run into any other troubles. If I try to shrink my world units any smaller, I start getting breakpoints hit in things like when adding a box shape with a width and height of 0.0140000004 . I guess this is too small for Jolt?

**Reply by SSNTails** [reply]
Aha! I have a repro for you! Check this out -- set the gravity in the sample app to 20, then load this snapshot (the layers are compatible this time). Look at how much things jitter and then easily clip through the static mesh.

[fall.zip](https://github.com/jrouwe/JoltPhysics/files/15042550/fall.zip)



https://github.com/jrouwe/JoltPhysics/assets/159929665/b9103bdd-11fd-4bcb-99ca-f9ac727ff117

**Reply by jrouwe** [reply, marked as answer]
It looks like your inertia calculations are off. If in `LoadSnapshotTest::Initialize` I add:

```
settings.mOverrideMassProperties = EOverrideMassProperties::CalculateInertia;
```

to the loop:

```
for (BodyCreationSettings &settings : scene->GetBodies())
```

Then it starts behaving normally.

You can visualize the inertia in the samples app by hitting the 'I' key, without the change the inertia looks like that of a tiny box:

![image](https://github.com/jrouwe/JoltPhysics/assets/1621693/84ce4d34-d828-454e-a7ad-0291f59a0857)

with the change it looks like a box that is about the same size as the barrel:

![image](https://github.com/jrouwe/JoltPhysics/assets/1621693/a6151b67-7a99-48b9-ab72-7603ad7e3de8)

When you use an inertia that is too small, small impulses lead to big angular velocity changes so it continuously bounces back and forth between both sides of the barrel as is visible in your video. As a general rule of thumb: Don't mess with inertia yourself, just let Jolt calculate it for you.

**Reply by SSNTails** [reply]
That was it!! Glad it was just some small setting I had wrong. I was doing this originally to supply a custom mass value:

bcs.mOverrideMassProperties = JPH::EOverrideMassProperties::MassAndInertiaProvided
bcs.mMassPropertiesOverride.SetMassAndInertiaOfSolidBox(JPH::Vec3(1, 1, 1), mass);

Doing as you said makes everything work:

bcs.mOverrideMassProperties = JPH::EOverrideMassProperties::CalculateInertia;
bcs.mMassPropertiesOverride.SetMassAndInertiaOfSolidBox(JPH::Vec3(1, 1, 1), mass);

**Reply by jrouwe** [reply]
Note that if you want to save a few cycles, instead of:

```
bcs.mMassPropertiesOverride.SetMassAndInertiaOfSolidBox(JPH::Vec3(1, 1, 1), mass);
```

you can do:

```
bcs.mMassPropertiesOverride.mMass = mass;
```

as the inertia will no longer be used.
